File: interviews/leetcode/leetcode-75/heap-and-priority-queue/2336. Smallest Number in Infinite Set/solution.py
class SmallestInfiniteSet:

    def __init__(self):
        self.added_back = set()  # Set to store the numbers that have been added back
        self.next_smallest = 1  # The next smallest number to be popped

    def addBack(self, n: int) -> None:
        # If the number is greater than or equal to next_smallest, no need to add it back
        if n < self.next_smallest:
            self.added_back.add(n)

    def popSmallest(self) -> int:
        if self.added_back:
            smallest = min(self.added_back)  # Get the smallest element in the added_back set
            self.added_back.remove(smallest)
        else:
            smallest = self.next_smallest  # Otherwise, take the next smallest number
            self.next_smallest += 1  # Increment the next smallest number
        return smallest

# Uncomment the following lines to test the function
# obj = SmallestInfiniteSet()
# obj.addBack(2)
# print(obj.popSmallest())  # Expected output: 1
# print(obj.popSmallest())  # Expected output: 2
# obj.addBack(1)
# print(obj.popSmallest())  # Expected output: 3
# print(obj.popSmallest())  # Expected output: 4

File: interviews/leetcode/leetcode-75/heap-and-priority-queue/2336. Smallest Number in Infinite Set/solution.go
package main

import (
	"math"
)

type SmallestInfiniteSet struct {
	addedBack    map[int]bool
	nextSmallest int
}

func Constructor() SmallestInfiniteSet {
	return SmallestInfiniteSet{
		addedBack:    make(map[int]bool),
		nextSmallest: 1,
	}
}

func (this *SmallestInfiniteSet) AddBack(n int) {
	// Add n back if it's smaller than the next smallest number
	if n < this.nextSmallest {
		this.addedBack[n] = true
	}
}

func (this *SmallestInfiniteSet) PopSmallest() int {
	var smallest int
	// If there are numbers that were added back, return the smallest one
	if len(this.addedBack) > 0 {
		smallest = math.MaxInt
		for num := range this.addedBack {
			if num < smallest {
				smallest = num
			}
		}
		delete(this.addedBack, smallest)
	} else {
		// Otherwise, return the next smallest number and increment it
		smallest = this.nextSmallest
		this.nextSmallest++
	}
	return smallest
}

// Uncomment the following lines to test the function
// obj := Constructor()
// obj.AddBack(2)
// fmt.Println(obj.PopSmallest())  // Expected output: 1
// fmt.Println(obj.PopSmallest())  // Expected output: 2
// obj.AddBack(1)
// fmt.Println(obj.PopSmallest())  // Expected output: 3
// fmt.Println(obj.PopSmallest())  // Expected output: 4
